catkin_tools/execution/stages.py

```diff
@@ -0,0 +1,83 @@
+
+import functools
+import os
+
+from .io import IOBufferContainer
+from .io import IOBufferLogger
+from .io import IOBufferProtocol
+
+
+class Stage(object):
+
+    """A description of one of the serially-executed stages of a Job.
+
+    Like Jobs, Stages are stateless, and simply describe what needs to be done
+    and how to do it.
+    """
+
+    def __init__(self, label, logger_factory=IOBufferProtocol.factory, occupy_job=True):
+        self.label = str(label)
+        self.logger_factory = logger_factory
+        self.occupy_job = occupy_job
+
+
+class CmdStage(Stage):
+
+    """Job stage that describes a system command.
+
+    :param label: The label for the stage
+    :param command: A list of strings composing a system command
+    :param protocol: A protocol class to use for this stage
+
+    Additional kwargs are passed to `async_execute_process`
+    """
+
+    def __init__(
+            self,
+            label,
+            cmd,
+            cwd=os.getcwd(),
+            env=None,
+            shell=False,
+            emulate_tty=True,
+            stderr_to_stdout=False,
+            occupy_job=True,
+            logger_factory=IOBufferProtocol.factory):
+        """ """
+
+        if not type(cmd) in [list, tuple] or not all([type(s) is str for s in cmd]):
+            raise ValueError('Command stage must be a list of strings: {}'.format(cmd))
+        super(CmdStage, self).__init__(label, logger_factory, occupy_job)
+
+        self.async_execute_process_kwargs = {
+            'cmd': cmd,
+            'cwd': cwd,
+            'env': env,
+            'shell': shell,
+            # Emulate tty for cli colors
+            'emulate_tty': emulate_tty,
+            # Capture stderr and stdout separately
+            'stderr_to_stdout': stderr_to_stdout
+        }
+
+
+class FunStage(Stage):
```

Comment:
Renamed `FunStage` to `FunctionStage` and `CmdStage` to `CommandStage` for consistency in 0f61d73